### PR TITLE
Cross-spork client checks for boundaries

### PR DESCRIPTION
**Pull request by sideninja**
Closes: #231 

## Description
This PR improves how cross-spork clients are handled. It adds a range for each client defined by first and last height in that spork, which enables more safety, first it enabled for all the clients to be checked if they create a continuous range of heights (without missing heights), secondly and more importantly, the requests error out if the height requested is out of range, which drops requests made to the AN and thus lower traffic and load.

Implementing this is important because there was a lot of requests for heights out of range being forwarded to ANs.

______

For contributor use:

- [ ] Targeted PR against `master` branch
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [standards mentioned here](https://github.com/onflow/flow-nft/blob/master/CONTRIBUTING.md#styleguides).
- [ ] Updated relevant documentation 
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Added appropriate labels 

## Summary by CodeRabbit

- **New Features**
  - Introduced cross-spork client capabilities for improved spork client management.

- **Improvements**
  - Enhanced error logging in blockchain functions for better debugging and user feedback.
  - Improved logging information for transactions and blocks.

- **Testing**
  - Refactored test setups with new utility functions for mock client configurations.
  - Added comprehensive tests for cross-spork client functionalities.

### CODE REVIEW

File: services/ingestion/subscriber_test.go [hunk 1]
```diff
@@ -8,79 +8,13 @@ import (
...
 
-	"github.com/onflow/flow-go-sdk"
-	"github.com/onflow/flow-go-sdk/access/mocks"
 	flowGo "github.com/onflow/flow-go/model/flow"
-	"github.com/onflow/flow-go/storage"
 	"github.com/rs/zerolog"
 	"github.com/stretchr/testify/require"
 )
 
-type mockClient struct {
```
Comment (sideninja) [inline services/ingestion/subscriber_test.go:20]: moved

File: services/testutils/mock_client.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
+package testutils
```
Comment (sideninja) [inline services/testutils/mock_client.go:1]: moved to here

File: services/requester/cross-spork_client.go [hunk 3]
```diff
@@ -2,16 +2,87 @@ package requester
...
+		lasts[i] = c.lastHeight
+	}
+
+	slices.Sort(firsts)
+	slices.Sort(lasts)
+
+	// make sure each last height is one smaller than next range first height
+	for i := 0; i < len(lasts)-1; i++ {
+		if lasts[i]+1 != firsts[i+1] {
+			return false
```
Comment (peterargue) [inline services/requester/cross-spork_client.go:79]: it'd be helpful to log which blocks were missing between which clients

File: services/requester/cross-spork_client.go [hunk 4]
```diff
@@ -2,16 +2,87 @@ package requester
...
+	return nil
+}
+
+// continuous checks if all the past spork clients create a continuous
+// range of heights.
+func (s *sporkClients) continuous() bool {
+	firsts := make([]uint64, len(*s))
+	lasts := make([]uint64, len(*s))
+
+	for i, c := range *s {
```
Comment (peterargue) [inline services/requester/cross-spork_client.go:68]: nit: I think you could simplify this logic by sorting the list of clients first, then iterating over the list once.

File: services/requester/requester.go [hunk 5]
```diff
@@ -236,24 +232,22 @@ func (e *EVM) signAndSend(ctx context.Context, script []byte, args ...cadence.Va
...
-				e.logger.Error().
-					Str("flow-id", id.String()).
-					Err(res.Error).
-					Msg("flow transaction failed to execute")
-				return
-			}
-
-			e.logger.Debug().
+	go func(id flow.Identifier) {
+		res, _ := e.client.GetTransactionResult(context.Background(), id)
```
[the remark this answers is not in the document]
Comment (sideninja) [reply services/requester/requester.go:236]: removed for https://github.com/onflow/flow-evm-gateway/issues/264
this change doesn't really belong in this PR

File: services/requester/cross-spork_client.go (hunk 3 above)
Comment (sideninja) [reply services/requester/cross-spork_client.go:79]: how would you report that? I guess changing this to return an error? but then what is the method? `mustBeContinuius` feels weird?

File: services/requester/cross-spork_client.go (hunk 4 above)
Comment (sideninja) [reply services/requester/cross-spork_client.go]: Good idea, added.

**Pull request merged**